clients/src/main/java/org/apache/kafka/clients/RecordHeaders.java

```diff
@@ -0,0 +1,177 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.kafka.common.utils.AbstractIterator;
+
+public class RecordHeaders implements Headers, Closeable {
+
+    private final List<Header> headers;
+    private volatile boolean isClosed = false;
+
+    public RecordHeaders() {
+        this(null);
+    }
+
+    public RecordHeaders(Iterable<Header> headers) {
+        if (headers == null) {
+            //new ArrayList, empty as no headers
+            this.headers = new ArrayList<>();
+        } else if (headers instanceof RecordHeaders) {
+            //new ArrayList using efficient Arrays.copyOf the backing list in given RecordHeaders
+            this.headers = new ArrayList<>(((RecordHeaders) headers).headers);
+        } else if (headers instanceof Collection) {
+            //new ArrayList using efficient Arrays.copyOf the given collection.
+            this.headers = new ArrayList<>((Collection<Header>) headers);
+        } else {
+            //new ArrayList iterating through to copy the headers.
+            this.headers = new ArrayList<>();
+            Iterator<Header> iterator = headers.iterator();
+            while (iterator.hasNext()) {
+                this.headers.add(iterator.next());
+            }
+        }
+    }
+
+    @Override
+    public Headers add(Header header) throws IllegalStateException {
+        if (isClosed) {
+            throw new IllegalStateException("RecordHeaders isClosed");
+        }
+        headers.add(header);
+        return this;
+    }
+
+    @Override
+    public Headers remove(String key) throws IllegalStateException {
+        if (isClosed) {
+            throw new IllegalStateException("RecordHeaders isClosed");
+        }
+        Iterator<Header> iterator = iterator();
+        while (iterator.hasNext()) {
+            if (iterator.next().key().equals(key)) {
+                iterator.remove();
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public Header lastHeader(String key) {
+        for (int i = headers.size() - 1; i >= 0; i--) {
```

Comment:
good point, behaviour should be as per other methods passing in a header key which is null, will update and add java doc.